### DISCUSSION TITLE
Graph shortest path Dijkstra - time optimisation

### DISCUSSION THREAD

**Discussion by ufraaan** [community/community#123513] [Programming Help]
# algorithm course progress
currently, I'm progressing through my Algorithm course.

## input data
the input data consists of two numbers:

- 𝑛 (2 ≤ 𝑛 ≤ 2⋅10^5): the number of settlements in the empire.
- 𝑚 (1 ≤ 𝑚 ≤ 4⋅10^5): the number of roads in it.

the following 𝑚 lines describe the roads between the settlements. each road is described by three numbers:

- 𝑣, 𝑢 (1 ≤ 𝑣, 𝑢 ≤ 𝑛, 𝑣 ≠ 𝑢): the settlements connected by the road.
- 𝑤 (1 ≤ 𝑤 ≤ 10^9): the length of the road.

## output
the output should contain:

- one natural number 𝑘 (2 ≤ 𝑘 ≤ 𝑛): the number of settlements on the shortest path from settlement 1 to settlement 𝑛.
- on the second line, separated by spaces, print the settlements on the shortest path from 1 to 𝑛 in traversal order, including 1 and 𝑛 themselves.
- if there is no path from 1 to 𝑛, output −1. In this case, you may begin to worry about the life of the coachman.
- if there are several answers, print any one.

time limit: 1 second  
memory limit per test: 256 megabytes

## python implementation
```python
import heapq
import sys

def dijkstra(graph, start, end):
    n = len(graph)
    distances = [float('inf')] * n
    distances[start] = 0
    prev = [-1] * n
    pq = [(0, start)]

    while pq:
        dist, u = heapq.heappop(pq)
        if u == end:
            path = [end]
            while prev[path[-1]] != -1:
                path.append(prev[path[-1]])
            path.reverse()
            return dist, path
        
        for v, w in graph[u]:
            if dist + w < distances[v]:
                distances[v] = dist + w
                prev[v] = u
                heapq.heappush(pq, (distances[v], v))

    return -1, []

# Input data
n, m = map(int, sys.stdin.readline().split())
graph = [[] for _ in range(n)]
for _ in range(m):
    v, u, w = map(int, sys.stdin.readline().split())
    graph[v-1].append((u-1, w))
    graph[u-1].append((v-1, w))

# Dijkstra function call
distance, path = dijkstra(graph, 0, n-1)

# Output
if distance == -1:
    sys.stdout.write("-1\n")
else:
    sys.stdout.write(str(len(path)) + "\n")
    sys.stdout.write(" ".join(str(x+1) for x in path) + "\n")
```
the algorithm works right, but i have problems with time limit. i have already made an optimization which works x2 faster from the one i have started with, but there are much more optimization ahead.

i have researched for about 10 days and read already a lot staff on this topic, but either i don't understand how to implement, or its just optimization

i have chosen [dijkstra algorithm](https://en.wikipedia.org/wiki/Dijkstra's_algorithm) as a classic way to solve this kind of problems.

after researching ways of realizations, I have chosen heap as the fastest recommended. Also tried another ways to collect the path.

also changed input output to sys.stdout and have some benefits at speed but still not enough * maybe there is a better algorithm for his task? or its just the problem of not proper optimization?

### guidelines

- [X] i have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.

**Comment by CopperEagle** [comment]
Hi @moroii69 

Hmm, I run the code for myself and it ran reasonably quick for large inputs. For $m = 5*10^5$ and $n = 10^4$ I get well below one second. Also, I think your choice for the algorithm valid. I'm just going to assume you are required to use Python - this code would definitely run faster using Java or  C++.

One thing that stood out to me was the excessive use of lists. If you have a lot of lookups, you should use tuples for this. They cannot be altered but  are faster. I used this technique for the graph by converting it to tuples after building the graph. However, the algorithm terminates so early (at least with random data) that the cost of converting is higher than the reduced cost of lookup. Maybe try this with the testcases your course gives you. It may be handcrafted to maximize the number of iterations.

**Comment by kazmatw** [comment]
Hi @moroii69 ✋
Here’s a minor optimization for reading input, which reduces the overhead of multiple I/O operations:

```python
import sys
import heapq

def dijkstra(graph, start, end):
    n = len(graph)
    distances = [float('inf')] * n
    distances[start] = 0
    prev = [-1] * n
    pq = [(0, start)]

    while pq:
        dist, u = heapq.heappop(pq)
        if u == end:
            path = [end]
            while prev[path[-1]] != -1:
                path.append(prev[path[-1]])
            path.reverse()
            return dist, path
        
        for v, w in graph[u]:
            if dist + w < distances[v]:
                distances[v] = dist + w
                prev[v] = u
                heapq.heappush(pq, (distances[v], v))

    return -1, []

input_data = sys.stdin.read().split()
index = 0
n, m = int(input_data[index]), int(input_data[index+1])
index += 2
graph = [[] for _ in range(n)]

for _ in range(m):
    v, u, w = int(input_data[index]), int(input_data[index+1]), int(input_data[index+2])
    graph[v-1].append((u-1, w))
    graph[u-1].append((v-1, w))
    index += 3

distance, path = dijkstra(graph, 0, n-1)

if distance == -1:
    sys.stdout.write("-1\n")
else:
    sys.stdout.write(str(len(path)) + "\n")
    sys.stdout.write(" ".join(str(x+1) for x in path) + "\n")
```
